### PR TITLE
test: add tests for sandbox creation rollback and cleanup

### PR DESCRIPTION
**Pull request by jingxiaolu**
When sandbox creation failed, we should do rollback operations to
cleanup those files and folders. Including:
- share path
- cgroup subsystems
- sandbox local resources

Fixes: #862

Signed-off-by: l00397676 <lujingxiao@huawei.com>

### CODE REVIEW

File: virtcontainers/container.go [hunk 1]
```diff
@@ -623,7 +623,10 @@ func newContainer(sandbox *Sandbox, contConfig ContainerConfig) (*Container, err
 // been performed before the container creation failed.
 // - Unplug CPU and memory resources from the VM.
 // - Unplug devices from the VM.
-func (c *Container) rollbackFailingContainerCreation() {
+func (c *Container) rollbackFailingContainerCreation(s *Sandbox) {
```
Comment (bergwolf) [inline virtcontainers/container.go]: You can always find `sandbox` in `c.sandbox`, so there is no need to pass in another sandbox as argument here.

File: virtcontainers/hyperstart_agent.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/kata-containers/runtime/virtcontainers/pkg/hyperstart"
 	ns "github.com/kata-containers/runtime/virtcontainers/pkg/nsenter"
 	"github.com/kata-containers/runtime/virtcontainers/utils"
-	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-spec/specs-go"
```
Comment (bergwolf) [inline virtcontainers/hyperstart_agent.go]: Won't it break build?

Comment (caoruidong) [reply virtcontainers/hyperstart_agent.go]: It won't. Better to leave it as is

Comment (jshachm) [reply virtcontainers/hyperstart_agent.go]: Maybe it's treat of `gland` IDE `gofmt` action. 

Leave it is better~

File: virtcontainers/container.go (hunk 1 above)
Comment (jingxiaolu) [reply virtcontainers/container.go]: code updated, thanks @bergwolf ~

File: virtcontainers/cgroups.go [hunk 3]
```diff
@@ -28,15 +28,15 @@ type sandboxCgroups struct {
 func (s *Sandbox) newCgroups() error {
 	// New will still succeed when cgroup exists
 	// create common parent for all kata-containers
-	// e.g. /sys/fs/cgroup/cpu/vc
+	// e.g. /sys/fs/cgroup/cpu/kata
```
Comment (sboeuf) [inline virtcontainers/cgroups.go]: Even if it's not a big deal, this change and the one below should not be part of this commit.

File: virtcontainers/container.go [hunk 4]
```diff
@@ -669,6 +672,7 @@ func createContainer(sandbox *Sandbox, contConfig ContainerConfig) (c *Container
 	defer func() {
 		if err != nil {
 			c.rollbackFailingContainerCreation()
+			c.sandbox.storage.deleteContainerResources(c.sandboxID, c.id, nil)
```
Comment (sboeuf) [inline virtcontainers/container.go]: There's a call to `c.removeResources()` from `rollbackFailingContainerCreation()`, so I would assume `c.sandbox.storage.deleteContainerResources()` is already part of `c.removeResources()`, right?
If it's not, make sure to add it to the rollback function instead of adding it here.

File: virtcontainers/agent.go [hunk 5]
```diff
@@ -155,6 +155,9 @@ type agent interface {
 	// createSandbox will tell the agent to perform necessary setup for a Sandbox.
 	createSandbox(sandbox *Sandbox) error
 
+	// cleanupSandbox will tell the agent to cleanup resources related to the Sandbox.
+	cleanupSandbox(sandbox *Sandbox) error
```
Comment (sboeuf) [inline virtcontainers/agent.go]: Since you're returning `nil` from all the implementations, which makes sense BTW, I would suggest this function not to return anything (void basically).

File: virtcontainers/sandbox.go [hunk 6]
```diff
@@ -758,6 +776,13 @@ func createSandbox(ctx context.Context, sandboxConfig SandboxConfig, factory Fac
 		return nil, err
 	}
 
+	defer func() {
+		// when fetchSandbox(), factory is nil, we should not release resources
+		if err != nil && factory != nil {
```
Comment (bergwolf) [inline virtcontainers/sandbox.go]: I'm not sure I get the reasoning. Except for `fetchSandbox()`, `factory` can still be nil when vm factory is disabled.

File: virtcontainers/sandbox.go [hunk 7]
```diff
@@ -740,6 +740,24 @@ func (s *Sandbox) getAndStoreGuestDetails() error {
...
 }
 
+func rollbackFailingSandboxCreation(ctx context.Context, s *Sandbox) {
+	span, _ := trace(ctx, "rollbackFailingSandboxCreation")
+	defer span.Finish()
+
+	if err := s.destroyCgroups(); err != nil {
+		s.Logger().WithError(err).Error("rollbackFailingSandboxCreation destroyCgroups failed")
+	}
+	if err := s.Release(); err != nil {
```
Comment (bergwolf) [inline virtcontainers/sandbox.go]: Please release sandbox last in case we want to access any hypervisor or agent resources although we don't right now.

File: virtcontainers/container.go (hunk 4 above)
Comment (jingxiaolu) [reply virtcontainers/container.go]: yes you're right, I moved it to `rollbackFailingContainerCreation()` now :+1:

File: virtcontainers/agent.go (hunk 5 above)
Comment (jingxiaolu) [reply virtcontainers/agent.go]: no `error` returns from `cleanupSandbox()` now~

File: virtcontainers/sandbox.go (hunk 6 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox.go]: :+1: important point! Now I move this part after comment `// Below code path is called only during create, because of earlier check.` Is it the right place now? :thinking:

File: virtcontainers/sandbox.go (hunk 7 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox.go]: corrected, pls check again~

**Comment by jingxiaolu** [comment]
Several tests added now, please help to check~ @sboeuf 

And I have a question... 
If I test `TestSandboxCreationRollback` with `KATA_GO_TEST_FLAGS='-v -run TestSandboxCreationRollback$'` only, it will `PASS`. 

But if I test it with all the other unit tests, it will report error `QEMU path (/tmp/vc-tmp-243276269/hypervisor) does not exist` in `newSandbox()` :cry: But actualy this `/tmp/vc-tmp-243276269/hypervisor` **was initialised** in `TestMain()` of `virtcontainers_test.go`

Do you have any suggestions? Many thanks!

File: virtcontainers/container.go [hunk 8]
```diff
@@ -633,6 +636,9 @@ func (c *Container) rollbackFailingContainerCreation() {
 	if err := c.removeDrive(); err != nil {
 		c.Logger().WithError(err).Error("rollback failed removeDrive()")
 	}
+	if err := c.sandbox.storage.deleteContainerResources(c.sandboxID, c.id, nil); err != nil {
+		c.Logger().WithError(err).Error("rollback failed storage deleteContainerResources()")
+	}
```
Comment (sboeuf) [inline virtcontainers/container.go]: The change here is related to the rollback of container creation, and it should be in its own commit. Please commit this separately (but in the same PR), and provide a nice commit message to explain what you did here.

File: virtcontainers/sandbox.go [hunk 9]
```diff
@@ -781,12 +799,18 @@ func createSandbox(ctx context.Context, sandboxConfig SandboxConfig, factory Fac
...
 
 	// Below code path is called only during create, because of earlier check.
-	if err := s.agent.createSandbox(s); err != nil {
+	defer func() {
+		if err != nil {
+			rollbackFailingSandboxCreation(ctx, s)
+		}
+	}()
+
+	if err = s.agent.createSandbox(s); err != nil {
```
Comment (sboeuf) [inline virtcontainers/sandbox.go]: `s.agent.createSandbox` should be called before since your rollback function includes the agent cleanup, right?

File: virtcontainers/sandbox.go [hunk 10]
```diff
@@ -781,12 +799,18 @@ func createSandbox(ctx context.Context, sandboxConfig SandboxConfig, factory Fac
 	}
 
 	// Below code path is called only during create, because of earlier check.
-	if err := s.agent.createSandbox(s); err != nil {
+	defer func() {
+		if err != nil {
+			rollbackFailingSandboxCreation(ctx, s)
```
Comment (sboeuf) [inline virtcontainers/sandbox.go]: You should not call this global function from here. It makes sense from `virtcontainers/api.go`, but here instead, we should have a finer granularity, by deferring a bunch of rollbacks after each function call succeeds. Something along those lines:
```go
if err = s.agent.createSandbox(s); err != nil {
	return nil, err
}

defer func() {
	if err != nil {
                s.agent.cleanupSandbox(s)
        }
}
```

File: virtcontainers/api.go [hunk 11]
```diff
@@ -74,6 +74,12 @@ func createSandboxFromConfig(ctx context.Context, sandboxConfig SandboxConfig, f
 		return nil, err
 	}
 
+	defer func() {
+		if err != nil {
+			rollbackFailingSandboxCreation(ctx, s)
```
Comment (sboeuf) [inline virtcontainers/api.go]: There's a problem with calling this function here. It contains a call deleting the sandbox resources, but they are created by `storeSandbox()`, which means they are not created yet at this point in the code.

File: virtcontainers/sandbox_test.go [hunk 12]
```diff
@@ -1748,3 +1749,82 @@ func TestStartNetworkMonitor(t *testing.T) {
...
+
+func checkDirNotExist(path string) error {
+	if _, err := os.Stat(path); os.IsExist(err) {
+		return fmt.Errorf("%s is still exists", path)
+	}
+	return nil
+}
+
+func TestSandboxCreationRollback(t *testing.T) {
+	defer cleanUp()
```
Comment (sboeuf) [inline virtcontainers/sandbox_test.go]: Do not defer this `cleanUp()`, just call before everything else in this function.

File: virtcontainers/api.go (hunk 11 above)
Comment (jingxiaolu) [reply virtcontainers/api.go]: Yes you're right. But what I'm trying to do is define a global rollback function and is put to `defer` at the very beginning of function. 

Or what you prefer is smaller rollback operations and adds after each operations?

Comment (sboeuf) [reply virtcontainers/api.go]: Because you're calling the defer right after `createSandbox()` succeeded, then we should expect to rollback only things that have been set up by `createSandbox()`.
So I guess to answer your question, several smaller rollbacks would be more appropriate.

Comment (jingxiaolu) [reply virtcontainers/api.go]: Updated according to your comments, please help to check again, thanks :smiley:

File: virtcontainers/sandbox.go (hunk 10 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox.go]: updated~

File: virtcontainers/sandbox.go (hunk 9 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox.go]: updated~

File: virtcontainers/container.go (hunk 8 above)
Comment (jingxiaolu) [reply virtcontainers/container.go]: This change is put to commit "sandbox: rollback container resources when create fail"~

File: virtcontainers/sandbox_test.go (hunk 12 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox_test.go]: updated~

File: virtcontainers/hyperstart_agent.go [hunk 13]
```diff
@@ -329,6 +329,20 @@ func (h *hyper) createSandbox(sandbox *Sandbox) (err error) {
 	return h.configure(sandbox.hypervisor, "", h.getSharePath(sandbox.id), false, nil)
 }
 
+func (h *hyper) cleanupSandbox(sandbox *Sandbox) {
+	if sandbox == nil || sandbox.id == "" {
+		h.Logger().Warn("sandbox is nil so exit on cleanup sandbox")
+		return
+	}
+
+	sharedPath := h.getSharePath(sandbox.id)
```
Comment (jshachm) [inline virtcontainers/hyperstart_agent.go]: do we have a chance that the shared dir is still mounted?

File: virtcontainers/api.go [hunk 14]
```diff
@@ -74,6 +74,14 @@ func createSandboxFromConfig(ctx context.Context, sandboxConfig SandboxConfig, f
 		return nil, err
 	}
 
+	// rollback the sandbox and cgroup subsystems
+	defer func() {
+		if err != nil {
+			s.deleteCgroups()
```
[the remark this answers is not in the document]
Comment (WeiZhang555) [reply virtcontainers/api.go]: It should be OK to keep it here, since `setupCgroups` is very last function in `createSandboxFromConfig`

File: virtcontainers/api.go [hunk 15]
```diff
@@ -109,13 +117,21 @@ func createSandboxFromConfig(ctx context.Context, sandboxConfig SandboxConfig, f
...
 
-	if err := s.getAndStoreGuestDetails(); err != nil {
+	if err = s.getAndStoreGuestDetails(); err != nil {
 		return nil, err
 	}
 
+	// rollback all sandbox related local filesystem storages
+	defer func() {
+		if err != nil {
+			s.storage.deleteSandboxResources(s.id, nil)
```
[the remark this answers is not in the document]
Comment (WeiZhang555) [reply virtcontainers/api.go]: Why not call `s.Delete()` ? It should consist everything to make a sandbox clean, if not, we should add more cleanup logics there.

File: virtcontainers/cgroups.go [hunk 16]
```diff
@@ -56,6 +56,15 @@ func (s *Sandbox) newCgroups() error {
 	return nil
 }
 
+func (s *Sandbox) deleteCgroups() error {
```
[the remark this answers is not in the document]
Comment (WeiZhang555) [reply virtcontainers/cgroups.go]: Yes, I don't see why this new function is necessary. `destroyCgroups` should work well for you

File: virtcontainers/hyperstart_agent.go (hunk 13 above)
Comment (jingxiaolu) [reply virtcontainers/hyperstart_agent.go]: wow, it seems we need an `EnsureRemoveAll()` to wrap `os.Removall()` to handle such scenario :face_with_head_bandage:

Comment (caoruidong) [reply virtcontainers/hyperstart_agent.go]: related: #912

File: virtcontainers/api.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jingxiaolu) [reply virtcontainers/api.go]: Yes, in `createSandbox() -> newSandbox()` we call `newCgroups()`, inside of it creates sub-cgroups in cgroup. So here we should `destroyCgroups()`

File: virtcontainers/api.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jingxiaolu) [reply virtcontainers/api.go]: @sboeuf I call `s.storage.deleteSandboxResources()` here because in `s.getAndStoreGuestDetails()` it calls `s.storage.storeSandboxResource(s.id, stateFileType, s.state)`

File: virtcontainers/cgroups.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (jingxiaolu) [reply virtcontainers/cgroups.go]: After discussed with @WeiZhang555 , I realised that `destroyCgroups()` is ok to use~ Code updated~

**Comment by jodh-intel** [comment]
Ping @jingxiaolu.

File: virtcontainers/sandbox_test.go [hunk 17]
```diff
@@ -1748,3 +1749,109 @@ func TestStartNetworkMonitor(t *testing.T) {
...
+	if err = checkDirNotExist(sandboxDirState); err != nil {
+		return fmt.Errorf("%s still exists", sandboxDirState)
+	}
+	if err = checkDirNotExist(configStoragePath); err != nil {
+		return fmt.Errorf("%s still exists", configStoragePath)
+	}
+	if err = checkDirNotExist(runStoragePath); err != nil {
+		return fmt.Errorf("%s still exists", runStoragePath)
+	}
+	if err = checkDirNotExist(path.Join("/sys/fs/cgroup/cpu/kata/", testSandboxID)); err != nil {
```
Comment (jodh-intel) [inline virtcontainers/sandbox_test.go]: It's generally not a good idea for tests to look at actual system files. I would create a variable (not a `const`) that the main code uses. That variable would build the path from `defaultCgroupParent`. Your test code can then redefine that variable so that the tests are not looking at `/sys`. For an example of what I'm talking about, see:

- https://github.com/kata-containers/runtime/blob/master/cli/kata-check_amd64_test.go#L52..L70

**Comment by jodh-intel** [comment]
@jingxiaolu - this branch has conflicts. Please could you resolve?

**Comment by jingxiaolu** [comment]
hi @jodh-intel thanks for reminding. I found @bergwolf have fixed this issue by #1058 and #1138 (thanks bro :stuck_out_tongue_closed_eyes: ), so two of my commits are not needed now:
- [sandbox: rollback sandbox resources when create fail](https://github.com/kata-containers/runtime/pull/880/commits/19b421a352afb4ea489f4a9873e0642521c0731d)
- [sandbox: rollback container resources when create fail](https://github.com/kata-containers/runtime/pull/880/commits/2e0e337141b80939b41ff8deba1cdda4f7fb8292)

I can rebase the test commit and push again, if it is still needed. Please share your comments~
- [test: add tests for sandbox creation rollback](https://github.com/kata-containers/runtime/pull/880/commits/19b421a352afb4ea489f4a9873e0642521c0731d)

**Comment by raravena80** [comment]
@jingxiaolu ping, any updates? Thx.

**Comment by jingxiaolu** [comment]
hi @jodh-intel @raravena80 now only tests commit is included in this pr, please help to review again! thanks~

File: virtcontainers/sandbox_test.go (hunk 17 above)
Comment (jingxiaolu) [reply virtcontainers/sandbox_test.go]: thanks for your comments, I found that this part of code not needed, so I deleted directly in new PR~

**Comment by nitkon** [comment]
Should we also cleanup when Sandbox Deletion fails?
Context: https://github.com/kata-containers/tests/issues/1157#issuecomment-464610446
@jingxiaolu @raravena80

**Comment by grahamwhaley** [comment]
/test
as the CIs are very red.
@jingxiaolu - is this stuck, or will you be working on it. I think we'd like to see the cleanups in kata :-)
@nitkon - ooi, is this at all realated to your #1247 at all - wonder if it makes sense for you to review here at all?

**Comment by jingxiaolu** [comment]
Hi @grahamwhaley this issue is trying to do cleanup when creation failed #862 . 
After a long time leave I found this issue is fixed as what I [mentioned before](https://github.com/kata-containers/runtime/pull/880#issuecomment-459581581). 
I think the tests commit ([test: add tests for sandbox creation rollback](https://github.com/kata-containers/runtime/pull/880/commits/19b421a352afb4ea489f4a9873e0642521c0731d)) is still needed?

The CI tests failed for no "Fixes" in the commit message, should I raise a new issue for the tests commit?
> Found 1 commit between commit HEAD and branch master
> ERROR: No "Fixes" found
> ERROR: checkcommits failed. See the document below for help on formatting
commits for the project.

**Comment by grahamwhaley** [comment]
Ah, I see @jingxiaolu - thanks for the clarificaton.
Yep, you'll need a Fixes. I just made an Issue for you - #1257 - please use that one :-)
Thanks for your patience!

**Comment by jingxiaolu** [comment]
@grahamwhaley Thanks a lot! Fixes is added to commit message and code rebased~

**Comment by grahamwhaley** [comment]
np @jingxiaolu - getting there - CI is now noting that one file needs a `gofmt`:

```
virtcontainers/hyperstart_agent_test.go:1::warning: file is not gofmted with -s (gofmt)
The command ".ci/static-checks.sh" failed and exited with 1 during .
```

**Comment by jingxiaolu** [comment]
@grahamwhaley static check is green now, but jenkins-ci-ubuntu-18-04 failed for 
> === RUN   TestVolumesSetFailingVoidArguments
> time="2019-02-23T06:16:18Z" level=error msg="Create new sandbox failed" error="QEMU path (/tmp/vc-tmp-372940724/hypervisor) does not exist" sandbox=7f49d00d-1995-4156-8c79-5f5ab24ce138 sandboxid=7f49d00d-1995-4156-8c79-5f5ab24ce138 source=virtcontainers subsystem=sandbox
> panic: test timed out after 30s

I think that's not because of my tests commit, could you help to retrigger the jenkins? Many thianks

**Comment by jingxiaolu** [comment]
wow, finally all jenkins are green now~ @grahamwhaley please help to review ~

**Comment by jshachm** [comment]
All pass~ let‘s get it merged~
